File: solutions/04.enterprise-bot/EnterpriseBotAccessors.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Threading;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;

namespace EnterpriseBot
{
    public class EnterpriseBotAccessors
    {
        public EnterpriseBotAccessors(UserState userState, ConversationState conversationState)
        {
            ConversationState = conversationState ?? throw new ArgumentNullException(nameof(conversationState));
            UserState = userState ?? throw new ArgumentNullException(nameof(userState));
        }

        /// <summary>
        /// Gets the <see cref="IStatePropertyAccessor{T}"/> name used for the <see cref="DialogState"/> accessor.
        /// </summary>
        /// <remarks>Accessors require a unique name.</remarks>
        /// <value>The accessor name for the <see cref="DialogStateProperty"/>accessor.</value>
        public static string DialogStateName { get; } = $"{nameof(EnterpriseBotAccessors)}.DialogState";

        /// <summary>
        /// Gets or sets the <see cref="IStatePropertyAccessor{T}"/> for <see cref="DialogState"/>.
        /// </summary>
        /// <value>
        /// The accessor stores the dialog state for the user.
        /// </value>
        public IStatePropertyAccessor<DialogState> DialogStateProperty { get; set; }

        /// <summary>
        /// Gets the <see cref="ConversationState"/> object for the conversation.
        /// </summary>
        /// <value>The <see cref="ConversationState"/> object.</value>
        public ConversationState ConversationState { get; }

        /// <summary>
        /// Gets the <see cref="UserState"/> object for the conversation.
        /// </summary>
        /// <value>The <see cref="UserState"/> object.</value>
        public UserState UserState { get; }
    }
}

File: samples/javascript_typescript/52.enterprise-bot/src/enterpriseBotAccessors.ts
import { UserState, ConversationState, StatePropertyAccessor } from 'botbuilder';
import { DialogState } from 'botbuilder-dialogs';

export class EnterpriseBotAccessors {
    private _userState: UserState;
    private _conversationState: ConversationState;
    private _dialogStateProperty!: StatePropertyAccessor<DialogState>;

    constructor(userState: UserState, conversationState: ConversationState) {
        if (!userState) throw ('Missing parameter.  userState is required'); 
        if (!conversationState) throw ('Missing parameter.  conversationState is required');
        this._userState = userState;
        this._conversationState = conversationState;
    }

    public static readonly DialogStateName: string = 'EnterpriseBotAccessors.DialogState';

    public get userState(): UserState {
        return this._userState;
    }

    public get conversationState(): ConversationState {
        return this._conversationState;
    }

    public get dialogStateProperty(): StatePropertyAccessor<DialogState> {
        return this._dialogStateProperty;
    }

    public set dialogStateProperty(dialogStateProperty: StatePropertyAccessor<DialogState>) {
        this._dialogStateProperty = dialogStateProperty;
    }
}